### ISSUE TITLE
Visual Studio 2022 Endpoint Explorer Integration

### ISSUE THREAD

**Issue by cmarkwick365** [ardalis/ApiEndpoints#196]
ApiEndPoints does not appear to work with VS 2022 Endpoint Explorer.  Is it possible to enable this feature or is there a way to get it working out of the box?

**Comment by ardalis** [comment]
Hmm, the feature is still in preview, right? 
https://devblogs.microsoft.com/visualstudio/web-api-development-in-visual-studio-2022/

I haven't used it myself but in the example shown in that article it's working with minimal API endpoints, though they claim it will also work with Controller-based endpoints. This library depends on Controllers, so any feature that discovers Controller-based endpoints should work with it. Do you have a sample you can share? Does the feature work with a vanilla controller class, but not with one that inherits from one of this package's base endpoint classes (which in turn inherit from controllers)?

**Comment by cmarkwick365** [comment]
Thanks for your fast response Steve.  I believe it's out of preview in VS 17.6,   I think it relies upon the [ApiController] attribute.
This article may help: 
https://dev.to/andytechdev/step-by-step-guide-testing-http-endpoints-in-visual-studio-2022-using-endpoints-explorer-fpb

**Comment by ardalis** [comment]
Hmm, well we have that attribute:
https://github.com/ardalis/ApiEndpoints/blob/main/src/Ardalis.ApiEndpoints/EndpointBase.cs#L8

**Comment by cmarkwick365** [comment]
I suspect it also utilizes the Http* attributes.

**Comment by ardalis** [comment]
I have VS 17.6.4 which says it's up to date and I can't find Endpoints Explorer anywhere. It's supposed to be here:
![image](https://github.com/ardalis/ApiEndpoints/assets/782127/88137aac-384d-4a10-bccb-510a27872fc2)

**Comment by ardalis** [comment]
Apparently the endpoints explorer menu option only appears if you first select a web project in solution explorer...

**Comment by emilun** [comment]
I can open the "Endpoints Explorer" window and it shows the api project. However, no endpoints are discovered.

**Comment by Edgaras91** [comment]
It may not be related, but for me, Endpoints Explorer only picks up if the controller class has `[ApiController]`, but it doesn't pick up if it inherits a custom base controller that has `[ApiController]` attribute:

WORKS:
```
    [ApiController]
    [Route("accounts")]
    public class AccountController : ApiControllerBase
```

DOESN'T WORK:
```
    [Route("accounts")]
    public class AccountController : ApiControllerBase
```
```
    [ApiController]
    [Route("api/[controller]")]
    public abstract class ApiControllerBase : ControllerBase
```

**Comment by ardalis** [comment]
Yeah, it definitely still has a ways to go to catch all of the endpoints in the project. They could also leverage the swagger file from the running project as a possible way to build their list of endpoints. Probably a good idea to leave feedback for them in their own repo and/or on social media where they'll see it (copy @visualstudio).

**Comment by johnholliday** [comment]
I've found that the Endpoints Explorer window only recognizes literal strings for routes.  For example, the following declaration shows up in the Endpoints Explorer window:

```csharp
public static class MyEndpoints {
  public static void MapMyEndpoints(this IEndpointRouteBuilder routes) {
    var group = routes.MapGroup("api/test");
    group.MapGet("/{id}", TestGetMethod);
}}
```

However, the following is invisible to the Endpoints Explorer window:

```csharp
public static class MyEndpoints {
  private const string ApiBase = "api/Test";
  public static void MapMyEndpoints(this IEndpointRouteBuilder routes) {
    var group = routes.MapGroup(ApiBase);
    group.MapGet($"{ApiBase}/{id}", TestGetMethod);
}}
```

**Comment by ardalis** [comment]
Definitely seems like a bug on their end. Have you reported it with the feedback icon in VS?

**Comment by ardalis** [comment]
Pinged them here:
https://twitter.com/ardalis/status/1701282105005646180

**Comment by johnholliday** [comment]
Thanks (although I’m not feeling very ‘X’ lately).  I did send feedback via VS, so let’s see.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by ardalis** [comment]
Yeah reports via this method are more likely to be acted upon, it seems:
https://learn.microsoft.com/en-us/visualstudio/ide/how-to-report-a-problem-with-visual-studio?view=vs-2022

**Comment by ramiabughazaleh** [comment]
Hi.

I found the Visual Studio feedback item @johnholliday reported regarding discovering non-literal strings for routes here:
https://developercommunity.visualstudio.com/t/Endpoints-Explorer-fails-to-recognize-no/10463837

This issue is still marked as "Under Consideration"

I found a Visual Studio feedback item related to the issue @cmarkwick365 and @Edgaras91 encountered, regarding `ApiControllerBase` class inheritance, and the `[ApiController]` attribute here:
https://developercommunity.visualstudio.com/t/Bug-in-Endpoints-Explorer:-Controllers-n/10359760

This issue is marked as "Fixed In: Visual Studio 2022 version 17.7".
